test/test_um2netcdf.py

```diff
@@ -5,7 +5,209 @@
+                     32908.69305496925, 39254.833576], dtype=object)
+    return data
+
+
+@pytest.fixture
+def mule_vars(z_sea_rho_data, z_sea_theta_data):
+    """Simulate mule variables from aiihca.paa1jan.subset data."""
+    d_lat = 1.25  # spacing manually copied from aiihca.paa1jan.subset file
+    d_lon = 1.875
+    return um2nc.MuleVars(um2nc.GRID_NEW_DYNAMICS, d_lat, d_lon, z_sea_rho_data, z_sea_theta_data)
+
+
+@pytest.fixture
+def air_temp_cube():
+    # copied from aiihca.paa1jan.subset file
+    return DummyCube(30204, "air_temperature")
+
+
+@pytest.fixture
+def precipitation_flux_cube():
+    # copied from aiihca.paa1jan.subset file
+    return DummyCube(5216, "precipitation_flux")
+
+
+@pytest.fixture
+def std_args():
+    # TODO: make args namedtuple?
+    args = mock.Mock()
+    args.nomask = False
+    args.nohist = False
+    args.nckind = 3
+    args.include_list = None
+    args.exclude_list = None
+    args.simple = False
+    args.verbose = False
+    return args
+
+
+def test_process_without_masking(air_temp_cube, precipitation_flux_cube, mule_vars, std_args):
+    """Attempts end-to-end test of process(), ignoring cubes requiring masking."""
+
+    # FIXME: this convoluted setup is a big code stench
+    #        use this to gradually refactor process()
+    #        add naming vars to the dummy cubes
+    with mock.patch("mule.load_umfile"):  # ignore m_load_umfile as process_mule_vars is mocked
+        with mock.patch("umpost.um2netcdf.process_mule_vars") as m_mule_vars:
+            m_mule_vars.return_value = mule_vars
+
+            with mock.patch("iris.load") as m_iris_load:
+                cubes = [air_temp_cube, precipitation_flux_cube]
+
+                for c in cubes:
+                    c.attributes = {um2nc.STASH: DummyStash(*split_item_code(c.item_code))}
+                    c.cell_methods = []
+
+                    # TODO: replace with DummyCubeWithCoords ()?
+                    #       or add coord fixtures?
+                    c.coord["latitude"] = 0.0  # FIXME
+                    c.coord["longitude"] = 0.0  # FIXME
+
+                m_iris_load.return_value = cubes
+
+                with mock.patch("iris.fileformats.netcdf.Saver") as m_saver:  # prevent I/O
+                    # mock `sman` var to prevent I/O
+                    m_sman = mock.Mock()
+                    m_saver().__enter__.return_value = m_sman
+
+                    # TODO: fix lat/lon & levels requires c.coord attribute
+                    #       use fixtures to add attrs & remove the patches?
+                    with mock.patch("umpost.um2netcdf.fix_latlon_coord") as m_coord:
+                        with mock.patch("umpost.um2netcdf.fix_level_coord") as m_level:
+                            with mock.patch("umpost.um2netcdf.apply_mask") as m_apply_mask:
+                                with mock.patch("umpost.um2netcdf.cubewrite") as m_cubewrite:
```

Comment:
Could we remove some of the nesting by having a multiple context manager? We're using `python >=3.10`, right? Could do something like this:
```py
with (
    mock.patch("umpost.um2netcdf.fix_latlon_coord") as m_coord,
    mock.patch("umpost.um2netcdf.fix_level_coord") as m_level,
    mock.patch("umpost.um2netcdf.apply_mask") as m_apply_mask,
    mock.patch("umpost.um2netcdf.cubewrite") as m_cubewrite,
):
    # usages of those contexts
```

Reply:
I'm new to the ACCESS-NRI ecosystem, can Python >= 3.10 be guaranteed? I seem to recall seeing some Python 3.9 references in other projects/systems...

Reply:
Issue #11 discusses our initial intent is to support Py 3.10+, although it's possible users require 3.8 or 3.9.
* Apparently the parenthesised context managers exist as a new feature in Py3.9: https://stackoverflow.com/questions/68924790/parenthesized-context-managers-work-in-python-3-9-but-not-3-8
* From 3.10+ we're safe with parenthesised context manager support:
https://docs.python.org/3.10/whatsnew/3.10.html#parenthesized-context-managers
* v3.9 is EOL in Oct 2025: https://devguide.python.org/versions/, so ~15 months off.

This leaves some options:
* Conservative deferral: upgrade to new ctx managers in October 2025
* Short term defer: assume 3.9 is safe with the new features, but defer upgrade to Oct 2024/3.8 EOL
* Fix now: assume baseline 3.10 adoption across the board

@aidanheerdegen would other stakeholders need to comment on their Py version requirements?

Reply:
Should get a comment from @jo-basevi. I recall the `payu` conda environments are Py3.9, but AFAICT that is because it is the default, and should be simple to specify a higher minimum version

https://github.com/ACCESS-NRI/payu-condaenv/blob/main/env.yml

There is an open issue in dev-docs about supported python versions, and I think we should have an organisation policy on what that should be, and a process to keep it updated.

https://github.com/ACCESS-NRI/dev-docs/issues/15

Reply:
I should have said above, I think we should shoot for Py3.10 as a guaranteed minimum python version.

Reply:
> Should get a comment from @jo-basevi. I recall the `payu` conda environments are Py3.9, but AFAICT that is because it is the default, and should be simple to specify a higher minimum version

I realised the parenthesised context manager block is limited to `test_um2netcdf.py`, **not** the workflow script. In which case, does `payu` ever run the unit tests for tools it calls in the workflow? (if not (& `gadi` envs are not an issue), we can probably treat the tests as 3.10+ & bask in modern Python joy...)

Reply:
I don't think payu runs unit tests for tools it calls in the workflow, so I think it's safe to use `3.10`+ for tests. Maybe newer versions of payu environments should move to using python 3.10

Reply:
As of the 6/8/2024 ESM1.5 standup, we'll support Python v3.10+ (see #11).

I'll do a few things:
* Reformat the unit tests
* Add a runtime warning if the minimum version isn't met (task in #11 )

The tests should also get cleaner & require less mocking as the modification functions are fixed up & can be called from `process()` by the tests.

Reply:
I've fixed the multiple nesting issue in that func with  https://github.com/ACCESS-NRI/um2nc-standalone/pull/58/commits/ec45532cc02ae1694e98804994a03a8d76e770a9.